apps/cnspec/cmd/backgroundjob/checkin.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package backgroundjob
+
+import (
+	"context"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/viper"
+	"go.mondoo.com/cnquery/v9"
+	"go.mondoo.com/cnquery/v9/cli/sysinfo"
+	"go.mondoo.com/cnquery/v9/providers-sdk/v1/upstream"
+	"go.mondoo.com/ranger-rpc"
+	"go.mondoo.com/ranger-rpc/plugins/scope"
+)
+
+type checkinPinger struct {
+	ctx        context.Context
+	interval   time.Duration
+	quit       chan struct{}
+	wg         sync.WaitGroup
+	endpoint   string
+	httpClient *http.Client
+	mrn        string
+	creds      *upstream.ServiceAccountCredentials
+}
+
+func NewCheckinPinger(ctx context.Context, httpClient *http.Client, endpoint string, config *upstream.UpstreamConfig, interval time.Duration) *checkinPinger {
+	return &checkinPinger{
+		ctx:        ctx,
+		interval:   interval,
+		quit:       make(chan struct{}),
+		endpoint:   endpoint,
+		httpClient: httpClient,
+		mrn:        config.Creds.Mrn,
+		creds:      config.Creds,
+	}
+}
+
+func (c *checkinPinger) Start() {
+	// determine information about the client
+	sysInfo, err := sysinfo.GatherSystemInfo()
+	if err != nil {
+		log.Error().Err(err).Msg("could not gather client information")
+		return
+	}
+	c.wg.Add(1)
+	runCheckIn := func() {
+		err := c.checkIn(sysInfo)
+		if err != nil {
+			log.Info().Err(err).Msg("could not perform check-in")
+		}
+	}
+
+	// run check-in once on startup
+	runCheckIn()
+
+	jitter := time.Duration(rand.Int63n(int64(c.interval)))
+	ticker := time.NewTicker(c.interval + jitter)
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-ticker.C:
+				runCheckIn()
+			case <-c.quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (c *checkinPinger) Stop() {
+	close(c.quit)
+	c.wg.Wait()
+}
+
+func (c *checkinPinger) checkIn(sysInfo *sysinfo.SystemInfo) error {
+	// gather service account
+	plugins := []ranger.ClientPlugin{}
+	plugins = append(plugins, sysInfoHeader(sysInfo, cnquery.DefaultFeatures))
+
+	credentials := c.creds
+	if credentials != nil && len(credentials.Mrn) > 0 {
+		certAuth, err := upstream.NewServiceAccountRangerPlugin(credentials)
+		if err != nil {
+			return errors.Wrap(err, "invalid credentials")
+		}
+		plugins = append(plugins, certAuth)
+	} else {
+		return errors.New("no credentials configured")
+	}
+
+	client, err := upstream.NewAgentManagerClient(c.endpoint, c.httpClient, plugins...)
+	if err != nil {
+		return errors.Wrap(err, "could not connect to mondoo platform")
+	}
+
+	name := viper.GetString("name")
```

Comment:
i think this isn't used anywhere at the moment

Reply:
Good catch. Removed it.